src/d3a/models/myco_matcher/myco_matcher.py

```diff
@@ -0,0 +1,55 @@
+"""
+Copyright 2018 Grid Singularity
+This file is part of D3A.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+from typing import Dict, Optional
+
+from d3a.d3a_core.util import is_external_matching_enabled
+from d3a.models.myco_matcher import MycoExternalMatcher, MycoInternalMatcher
+from d3a.models.myco_matcher.myco_matcher_interface import MycoMatcherInterface
+
+
+class MycoMatcher(MycoMatcherInterface):
```

Comment:
I needed some time and re-reading your PR description to understand was this class does.
IMO a class name like the following would make it more clear right away. But I may be the only one. Open for other proposals too.
```suggestion
class MycoMatcherBase(MycoMatcherInterface):
```

Reply:
IMO `Base` could mean that this is a base class that others will extend, which is not the case here as this is more like a `wrapper` for `MycoInternalMatcher` and `MycoExternalMatcher`
What do you think about `MycoMatchersWrapper`?

Reply:
It is indeed a wrapper, I was confused by the fact that all three classes inherit from `MycoMatcherInterface` . However, l would not vote for adding "Wrapper" to the name. Leave it as is or somebody else has a better idea.

Reply:
The structure looks like a sort of [Bridge Pattern](https://refactoring.guru/design-patterns/bridge/python/example). Maybe we could change the names of the implementation classes (e.g. `MycoExternalMatcherImplementation`) and not the name of this `MycoMatcher` class.

---

One of the differences with the bridge pattern is that in our case it's not currently possible to do something like this:
```python
matcher_implementation = MycoExternalMatcher()
MycoMatcher(implementation=matcher_implementation)
```
(That's because we want to instantiate this as a singleton before we know its implementation).

Reply:
@fievelk Thanks for the addition!
I thought about conditionally passing the selected `implementation` to the Myco matcher in the singleton definition and as you correctly stated this can't work because this is called even before the configurations are updated
Honestly we're not interested in having only one MycoMatcher instance, we're only interested in the one in the `singletons` module .
Although it doesn't make sense to have 2 Myco Matchers, there is no harm if this happens, if we reassign this variable the python gc should remove it anyway and if we create an instance in other places it won't bother us (this is not like instantiating two different connections instances for example)
Anw the current implementation makes it easy to alter the matcher (implementation) or matcher's properties (matching algorithm) in the runtime, for example we have the `activate` method and we could introduce another `reactivate-update or any other naming`
One extra benefit from wrapping Internal and External matchers is that we have full control flow, it's true both have the same interface but the callbacks are a bit different (External matcher's match_recommendations is called upon the arrive of a redis message)
Another example is the `event_tick`, for internal matchers we always want to forward the tick whereas for the external matchers we only forward a percentage of the ticks.

Anyway the naming `Implementation` sounds good to me if you both agree

Reply:
> Honestly we're not interested in having only one MycoMatcher instance, we're only interested in the one in the singletons module .

I wasn't suggesting to change the `MycoMatcher` class to turn it into a singleton. I was suggesting to create a `SingletonBidOfferMatcher` class (or any name) that would wrap the `MycoMatcher` and could be used as a singleton. In this way, the `MycoMatcher` could be instantiated with the implementation parameter. Anyway, this is not so important here.

Regarding the different implementation of `MycoExternal/InternalMatcher`, I'm not completely sure I understood your point. The implementation is of course different, but they can be used interchangeably just by setting them as `MycoMatcher().matcher`, right? If not, something might not be quite right, because we should be able to call the same methods on `MycoMatcher().matcher` regardless of the object that this variable contains (without using `if`s).

Reply:
Of course they both have the same interfaces `event_tick` `event_market_cycle` `event_finish` `match_recommendations`, it's just that as I mentioned the example of dispatching the tick event, we don't want to dispatch **All** ticks to the External matcher so we have a check here, the other exception is that we do not want to call the `match_recommendations` of the external matcher from within the area tick event (we call it once we receive a redis command)

Reply:
I personally think this could be improved: we should not differentiate based on the content of `MycoMatcher().matcher`, otherwise this defeats the point of using a common interface.
For example, these lines:
```python
if (self.simulation_config.external_connection_enabled and
        is_external_matching_enabled()):
    # If External matching is enabled, limit the number of ticks dispatched.
    if external_global_statistics.is_it_time_for_external_tick(
            current_tick_in_slot):
        bid_offer_matcher.event_tick()
else:
    bid_offer_matcher.event_tick()
```
IMO should become:
```python
bid_offer_matcher.event_tick()
```
Then you would apply the `if` statements inside the `event_tick()` implementation.

Same reasoning goes for the `match_recommendations()` call (I couldn't find the point you were mentioning).

Reply:
BTW, this is the exact same thing that Hannes suggested in his comment https://github.com/gridsingularity/d3a/pull/1197#pullrequestreview-722358911

Reply:
@fievelk Hannes initially suggested moving this check from within the simulation.py (caller code)
So I moved it inside the MycoMatcher, one major difference from the Bridge to notice is that in the MycoMatcher I also extend the interface and forwarding messages inside the matchers after having my own control logic in there 
This way, the caller code doesn't call `bid_offer_matcher.matcher.event_tick()`, it is enough to do `bid_offer_matcher.event_tick()` .

Reply:
I took some time to understand the Bridge Pattern but I fail to see the connection to our "Problem" here. 
I would agree that selecting the Mycomatcher already at initiation of the MycoMatcher class and not only when calling the activate method (why do we need this two-stage init here anyway?).
Apart from that I am having trouble to follow the discussion here. Could maybe somebody write a short summary what the discussion solutions are? Thanks

Reply:
I have adapted to the suggestions and removed the MycoMatcherInterface from the Myco matcher, this way it only serves as a bridge without having any logic control flow, let me know what you think

Reply:
@hannesdiedrich we have a two stage init here because the get_setup() methods of simulations set the SpotMarketType and MatchingAlgorithm at runtime, this happens after the initialization of the Myco Matcher so in this case the myco matcher will always be initialized with the default settings (Internal matching with PAB )
This is why there is an activate method that will be called after the settings are loaded

Reply:
Bridge pattern:
> The abstraction will be able to delegate some (sometimes, most) of its calls to the implementations object. Since all implementations will have a common interface, they’d be interchangeable inside the abstraction.

In our case I considered the "abstraction" to be `MycoMatcher`, and the implementations are `MycoExternalMatcher` and `MycoInternalMatcher`.
The discussion was about the fact that our implementations are not interchangeable (we couldn't just call `bid_offer_matcher.matcher.event_tick()` without adding `if` statements or custom function parameters).

Reply:
Thanks a lot for this summary! So if my comment https://github.com/gridsingularity/d3a/pull/1197#discussion_r683374174 would be addressed, the bridge pattern could be applied and we decide which implementation is used in simulation.init and not in the Abstraction class, right?

Reply:
The bridge pattern is already implemented, in the Myco Matcher we have a property `matcher`, so in the future we can create more matchers and still make sure they can fit inside the MycoMatcher and still the interface won't break
```python
class NewMatcher(MycoMatcherInterface):
# new way of doing things
```
In the MycoMatcher:
```python
        if is_external_matching_enabled():
            return MycoExternalMatcher()
        elif conditionX:
            return MycoInternalMatcher()
       else:
           return NewMatcher() # This is it, the codebase would still be working exactly as expected
```

Reply:
To get closer to a real Bridge patter we would need to avoid calling `bid_offer_matcher.matcher`: the `.matcher` should be a private variable (e.g. `_matcher` or `_implementation`) that is not seen from the outside, and `MycoMatcher` should provide just one unique interface that interact internally with `_matcher`.

> Thanks a lot for this summary! So if my comment #1197 (comment) would be addressed, the bridge pattern could be applied and we decide which implementation is used in simulation.init and not in the Abstraction class, right?

@Hassan754 I think that @hannesdiedrich means that it would be better to just pass the matcher implementation to the `MycoMatcher` init:
```python
# simulation.py
matcher_implementation = MycoExternalMatcher() if is_external_matching_enabled() else MycoInternalMatcher()
bid_offer_matcher = MycoMatcher(implementation=matcher_implementation)
```
I would personally vote for this, because your current solution (adding new `else` statements inside the `MycoMatcher`) break the Open/Closed SOLID principle:
> write your code so that you will be able to add new functionality without changing the existing code

To be honest, the choice of which myco matcher implementation to use should be made by yet another class if we want to be stricter :D

Reply:
Indeed the last piece of separating the abstractions from the implementation is by defining MycoMatcher's unique interface, however AFAIS there are some places in the code we're accessing matcher.matching_algorithm.state `src/d3a/d3a_core/export.py:219` so I decided to keep the `matcher` property public ( However I very much believe that this implementation should be changed to make the design complete, but I decided not to interfere with these -at least now- as Faizan is in the same territory fixing the PAC exports and we're way out of the context of the ticket)

For the second point, yes it is not feasible due to the `activate` and possibly the option to change the matchers/matching algorithm in the runtime (as Events for example)
What I tried to do is the `get_matcher` static_method as a workaround to evitate creating an extra class, but if u believe this is inevitable, I'll create it . Thanks again for bearing with me !

Reply:
> What I tried to do is the get_matcher static_method as a workaround to evitate creating an extra class, but if u believe this is inevitable, I'll create it . Thanks again for bearing with me !

I don't think that right now this is necessary. I mentioned it because in your example you were adding another `else` to this method, and in that case I think it would make sense to have some separate logic for it. No need to do it now.